### PR TITLE
Implement CI.

### PR DESCRIPTION
**Pull request by genius3000**
Limited to a push or pull request that changes anything in the 3.0 directory. This could maybe be improved at some point, e.g., for the next version's modules to build against that branch as well.

Installs any additional dependencies (modules should include the PackageInfo directive if necessary).

Builds are based on the inspircd/inspircd insp3 branch and only build the necessary subset of modules:
- Pull Request: build any modules within the 3.0 directory that have been changed.
- Push: build all modules within the 3.0 directory.

### CODE REVIEW

**Comment by SadieCat** [comment]
Is there a specific need to have `tools/test-build` for this? That's mainly used in inspircd/inspircd to allow us to test socket engines but modules should not be affected by this.

Instead, we can just do something like this:

```sh
make `INSPIRCD_TARGET="m_foo m_bar m_baz"
```

**Comment by genius3000** [comment]
No, I was just mimicking the test of inspircd/inspircd; slimming it down here though would be good for sure.

File: .github/workflows/ci-macos.yml [hunk 1]
```diff
@@ -36,12 +38,15 @@ jobs:
...
       - name: Install additional dependencies
         if: env.PACKAGES != ''
         run: brew install ${{ env.PACKAGES }}
-      - name: Run test-build-contrib
-        run: cd inspircd; ./tools/test-build-contrib c++
+      - name: Build
+        run: |
+          cd inspircd
+          ./configure --development --disable-auto-extras
+          make --jobs $(($(nproc)+1))
```
Comment (SadieCat) [inline .github/workflows/ci-macos.yml:49]:
```suggestion
          make --jobs $(($(getconf _NPROCESSORS_ONLN) + 1))
```

File: .github/workflows/ci-macos.yml [hunk 2]
```diff
@@ -11,6 +11,9 @@ jobs:
     runs-on: macos-latest
     env:
       CXXFLAGS: -std=${{ matrix.standard }} -Wno-error=deprecated-declarations
+      CXX: c++
```
Comment (SadieCat) [inline .github/workflows/ci-macos.yml:14]: There's no need to set this as its the default.

File: .github/workflows/ci-linux.yml [hunk 3]
```diff
@@ -24,23 +27,25 @@ jobs:
...
       - name: Install additional dependencies
         if: env.PACKAGES != ''
         run: sudo apt-get install --assume-yes --no-install-recommends ${{ env.PACKAGES }}
-      - name: Run test-build-contrib
-        run: cd inspircd; ./tools/test-build-contrib ${{ matrix.compiler }}
+      - name: Build
+        run: |
+          cd inspircd
+          ./configure --development --disable-auto-extras
+          make --jobs $(($(nproc)+1))
```
Comment (SadieCat) [inline .github/workflows/ci-linux.yml:48]:
```suggestion
          make --jobs $(($(getconf _NPROCESSORS_ONLN) + 1))
```

File: .github/workflows/ci-macos.yml [hunk 4]
```diff
@@ -36,12 +38,15 @@ jobs:
...
         run: |
           cd inspircd
-          echo ::set-env name=PACKAGES::$(for f in src/modules/m_*.cpp; do ./tools/directive $f PackageInfo; done)
+          echo ::set-env name=PACKAGES::$(for f in ../contrib/3.0/*; do ./tools/directive $f PackageInfo; done)
       - name: Install additional dependencies
         if: env.PACKAGES != ''
         run: brew install ${{ env.PACKAGES }}
-      - name: Run test-build-contrib
-        run: cd inspircd; ./tools/test-build-contrib c++
+      - name: Build
```
Comment (SadieCat) [inline .github/workflows/ci-macos.yml:45]:
```suggestion
      - name: Build the modules
```

**Comment by SadieCat** [comment]
I wonder if it would be possible to get a list of changed modules from Git and only rebuild those?

**Comment by genius3000** [comment]
I don't think the event supplies that info but I had seen somewhere a mention of diff'ing the whole tree by revision (or such).

I had wanted to do a full check always, but thinking about it, it would suck to have a build fail for totally unrelated reasons to one's change(s). I'll have another look at all the events we can trigger on, maybe we can trigger off the base repo or just use an interval. That way we can still do a full CI incase something in the base repo causes a breakage in an untouched module here.

**Comment by SadieCat** [comment]
It's fine to do it this way for now. In the future we can look into changing it if we decide to.

**Comment by SadieCat** [comment]
LGTM.

**Comment by genius3000** [comment]
@SadieCat I squashed the 'build diet' and 'resolving suggestions' commits into the base 'Implement' commit. I'd appreciate another review on the commit for the PR and Push event separation. I did run it through a full PR cycle in my fork; opening a PR runs the PR check, merging a PR runs the push check.

**Pull request merged**